File: Algorithms/leetcode/Ex_0022_Generate_Parentheses/Soln_4_Bfs.cpp
#include <string>
#include <vector>
#include <queue>

using std::string;
using std::vector;
using std::queue;

class Node {
public: 
    // 当前得到的字符串
    string str;
    // 剩余左括号数量
    int left;
    // 剩余右括号数量
    int right;

    Node(string s, int l, int r) : str {std::move(s)}, left {l}, right{r} {
    }
};

class Solution_4 {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        if (n == 0) {
            return res;
        }
        queue<Node> queue;
        queue.push(Node("", n, n));
        while (!queue.empty()) {
            Node curNode = queue.front();
            queue.pop();
            if (curNode.left == 0 && curNode.right == 0) {
                // 结算
                res.push_back(curNode.str);
            }
            if (curNode.left > 0) {
                queue.push(Node(curNode.str + "(", curNode.left - 1, curNode.right));
            }
            if (curNode.right > 0 && curNode.left < curNode.right) {
                // 只有左括号剩余比右括号少时（即左括号使用的比右括号多时），才能添加右括号
                queue.push(Node(curNode.str + ")", curNode.left, curNode.right - 1));
            }
        }
        return res;
    }
};

// 参考题解：https://leetcode-cn.com/problems/generate-parentheses/solution/hui-su-suan-fa-by-liweiwei1419/

File: Algorithms/leetcode/Ex_0297_Serialize_and_Deserialize_Binary_Tree/Soln_4_Bfs.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    const queue = [root];
    let res = [];
    while (queue.length) {
        const node = queue.shift(); // 弹出队头节点
        if (node) {
            res.push(node.val);
            queue.push(node.left);
            queue.push(node.right);
        } else {
            res.push('X');
        }
    }
    return res.join(',');
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    if (data == 'X') {
        return null;
    }
    const list = data.split(',');
    const root = new TreeNode(list[0]);
    const queue = [root];
    let index = 1;
    while (index < list.length) {
        const node = queue.shift(); // 弹出队头节点
        const leftVal = list[index];
        index += 1;
        const rightVal = list[index];
        index += 1;

        if (leftVal != 'X') {
            const leftNode = new TreeNode(leftVal);
            node.left = leftNode;
            queue.push(leftNode);
        }

        if (rightVal != 'X') {
            const rightNode = new TreeNode(rightVal);
            node.right = rightNode;
            queue.push(rightNode);
        }
    }
    return root;
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

// 参考题解（强烈推荐）：https://leetcode-cn.com/problems/serialize-and-deserialize-binary-tree/solution/shou-hui-tu-jie-gei-chu-dfshe-bfsliang-chong-jie-f/
